File: src/main/java/com/tobe/healthy/schedule/domain/entity/TrainerScheduleClosedDaysInfo.java
package com.tobe.healthy.schedule.domain.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.DynamicUpdate;

import java.time.DayOfWeek;

@Entity
@Getter
@DynamicUpdate
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class TrainerScheduleClosedDaysInfo {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "trainer_schedule_closed_days_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "trainer_schedule_info_id")
    private TrainerScheduleInfo trainerScheduleInfo;

    @Enumerated(EnumType.STRING)
    private DayOfWeek closedDays;

    private TrainerScheduleClosedDaysInfo(TrainerScheduleInfo trainerScheduleInfo, DayOfWeek closedDays) {
        this.trainerScheduleInfo = trainerScheduleInfo;
        this.closedDays = closedDays;
    }

    public static TrainerScheduleClosedDaysInfo registerClosedDay(
            DayOfWeek dayOfWeek,
            TrainerScheduleInfo trainerScheduleInfo
    ) {
        return new TrainerScheduleClosedDaysInfo(trainerScheduleInfo, dayOfWeek);
    }
}

File: src/main/kotlin/com/tobe/healthy/schedule/domain/entity/TrainerScheduleClosedDaysInfo.kt
package com.tobe.healthy.schedule.domain.entity

import jakarta.persistence.*
import jakarta.persistence.EnumType.STRING
import jakarta.persistence.FetchType.LAZY
import jakarta.persistence.GenerationType.IDENTITY
import org.hibernate.annotations.DynamicUpdate
import java.time.DayOfWeek

@Entity
@DynamicUpdate
class TrainerScheduleClosedDaysInfo(

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "trainer_schedule_info_id")
    val trainerScheduleInfo: TrainerScheduleInfo? = null,

    @Enumerated(STRING)
    var closedDays: DayOfWeek,

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "trainer_schedule_closed_days_id")
    val id: Long? = null
) {

    companion object {
        fun registerClosedDay(
            dayOfWeek: DayOfWeek,
            trainerScheduleInfo: TrainerScheduleInfo
        ): TrainerScheduleClosedDaysInfo {
            return TrainerScheduleClosedDaysInfo(
                trainerScheduleInfo = trainerScheduleInfo,
                closedDays = dayOfWeek
            )
        }
    }
}
